---
title: "Fig 6 - RNA-sequencing"
author: "Benjamin Ostendorf"
date: "2023/04/24"
output: rmarkdown::github_document
---

## Preamble
```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(tidyverse)

## Color palette
cols_RGX <- c("ctrl" = "#A9A9A8", "tx" = "#0095FF")
```

## Import data
```{r, message=FALSE, warning=FALSE}
## Assemble dds object
read_counts <- read.table("counts.tsv", header = TRUE, row.names = "gene")

condition <- factor(c(rep("ctrl", 4), rep("tx", 4)), levels = c("ctrl", "tx"))
sample_info <- data.frame(condition = condition, 
                          row.names = names(read_counts))

dds <- DESeqDataSetFromMatrix(countData = read_counts,
                              colData = sample_info,
                              design = ~ condition)

## Remove uninformative columns
dds <- dds[rowSums(counts(dds)) > 1, ]

## Normalization and preprocessing
dds <- DESeq(dds)

## Get log2-transformed and scaled gene expression data
dds_rlog <- rlog(dds)
counts_norm <- counts(dds, normalized = TRUE)
counts_log2 <- log2(counts_norm + 1)
counts_scaled <- t(scale(t(counts_log2)))

## Get shrunken lfc results
res <- results(dds, contrast = c("condition", "tx", "ctrl"))
resLFC <- 
  lfcShrink(dds, res = res, contrast = c("condition", "tx", "ctrl"), type = "ashr") |>
  as_tibble(rownames = "symbol") |>
  arrange(padj)

write_tsv(resLFC, "resLFC_RNAseq.tsv")
counts_scaled |>
  as_tibble(rownames = "symbol") |>
  write_tsv("counts_scaled_RNAseq.tsv")
```

## Enrichment analysis with Clusterprofiler for GO
```{r, message=FALSE, warning=FALSE}
resLFC$entrez <- 
  mapIds(org.Mm.eg.db, 
         keys = resLFC$symbol, 
         keytype = "SYMBOL", 
         column = "ENTREZID")

genes_GO_entrez <- 
  resLFC |>
  filter(!is.na(padj) & padj < 0.1) |>
  filter(log2FoldChange > 0) |>
  pull(entrez)

ego <- enrichGO(gene          = genes_GO_entrez,
                universe      = resLFC$entrez,
                OrgDb         = org.Mm.eg.db,
                ont           = c("BP"), 
                pAdjustMethod = "fdr",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.1)

ego_simp <- clusterProfiler::simplify(ego, cutoff = 0.6, by = "p.adjust", select_fun = min)
ego_simp <- setReadable(ego_simp, 'org.Mm.eg.db', 'ENTREZID')

results_GO <- summary(ego_simp)

theme_dotplot <- 
  theme(
    axis.title = element_text(size = 6), 
    axis.text.x = element_text(size = 5), 
    axis.text.y = element_text(size = 5), 
    legend.text = element_text(size = 5), 
    legend.title = element_text(size = 5)
  )

GO_dotplot <- 
  dotplot(ego_simp, 
          showCategory = 15) +
  theme_dotplot +
  scale_size_continuous(range = c(0.5, 2)) +
  theme(legend.key.size = unit(.1, "cm"), 
        legend.position = "right", 
        axis.text.x = element_text(size = 4))

print(GO_dotplot)
```

## Session info
```{r}
devtools::session_info()
```